From 4edd47c070ce619e1029a0dc7a386d79c23ee05a Mon Sep 17 00:00:00 2001
From: Jan Horstmann <horstmann@osism.tech>
Date: Mon, 13 Jan 2025 11:29:36 +0100
Subject: [PATCH] Add graceful shutdown to ovn-controller unit

Add systemd service override to trigger a graceful shutdown of the
ovn-controller, so that it can deregister its chassis from OVN.
Additionally introduce ordering to ensure that ovn-controller is
shutdown before OVS DB and vSwitch services, so that the chassis may be
deregistered properly.

Closes-Bug: #2093328

Change-Id: I3bc6ce3599df6d0c208baf2ceee10c223471ca49
Signed-off-by: Jan Horstmann <horstmann@osism.tech>
---
 .../roles/ovn-controller/handlers/main.yml    |  5 ++++
 ansible/roles/ovn-controller/tasks/config.yml | 23 +++++++++++++++++++
 .../ovn-controller-systemd-override.conf.j2   | 14 +++++++++++
 .../notes/bug-2093328-db7299bb092fb12b.yaml   |  8 +++++++
 4 files changed, 50 insertions(+)
 create mode 100644 ansible/roles/ovn-controller/templates/ovn-controller-systemd-override.conf.j2
 create mode 100644 releasenotes/notes/bug-2093328-db7299bb092fb12b.yaml

diff --git a/ansible/roles/ovn-controller/handlers/main.yml b/ansible/roles/ovn-controller/handlers/main.yml
index 6068d9fb4..874f9ed76 100644
--- a/ansible/roles/ovn-controller/handlers/main.yml
+++ b/ansible/roles/ovn-controller/handlers/main.yml
@@ -1,4 +1,9 @@
 ---
+- name: Reload systemd config
+  become: true
+  ansible.builtin.systemd_service:
+    daemon_reload: true
+
 - name: Restart ovn-controller container
   vars:
     service_name: "ovn-controller"
diff --git a/ansible/roles/ovn-controller/tasks/config.yml b/ansible/roles/ovn-controller/tasks/config.yml
index 62a37a945..621131b0e 100644
--- a/ansible/roles/ovn-controller/tasks/config.yml
+++ b/ansible/roles/ovn-controller/tasks/config.yml
@@ -18,3 +18,26 @@
   with_dict: "{{ ovn_controller_services | select_services_enabled_and_mapped_to_host }}"
   notify:
     - Restart {{ item.key }} container
+
+- name: Ensuring systemd override directory exists
+  vars:
+    service_name: "kolla-{{ item.key | replace('-', '_') }}-container.service"
+  file:
+    path: "/etc/systemd/system/{{ service_name }}.d"
+    state: "directory"
+    mode: "0770"
+  become: true
+  with_dict: "{{ ovn_controller_services | select_services_enabled_and_mapped_to_host }}"
+  when: item.key == 'ovn-controller'
+
+- name: Copying over systemd override
+  vars:
+    service_name: "kolla-{{ item.key | replace('-', '_') }}-container.service"
+  template:
+    src: "{{ item.key }}-systemd-override.conf.j2"
+    dest: "/etc/systemd/system/{{ service_name }}.d/override.conf"
+    mode: "0660"
+  become: true
+  with_dict: "{{ ovn_controller_services | select_services_enabled_and_mapped_to_host }}"
+  when: item.key == 'ovn-controller'
+  notify: Reload systemd config
diff --git a/ansible/roles/ovn-controller/templates/ovn-controller-systemd-override.conf.j2 b/ansible/roles/ovn-controller/templates/ovn-controller-systemd-override.conf.j2
new file mode 100644
index 000000000..84304d6c5
--- /dev/null
+++ b/ansible/roles/ovn-controller/templates/ovn-controller-systemd-override.conf.j2
@@ -0,0 +1,14 @@
+# /etc/systemd/system/{{ service_name }}
+# {{ service_name }}
+# autogenerated by Kolla-Ansible
+
+[Unit]
+# NOTE(jhorstmann): ovn-controller interacts with OVS DB and vSwitch
+# Ensure that both are still available during graceful shutdown
+After=kolla-openvswitch_vswitchd.service kolla-openvswitch_db.service
+
+[Service]
+# NOTE(jhorstmann): Override stop command to include graceful shutdown of ovn-controller
+ExecStop=
+ExecStop=-/usr/bin/{{ kolla_container_engine }} exec ovn_controller /bin/sh -c 'ovs-appctl --target /run/ovn/ovn-controller.*.ctl exit'
+ExecStop=/usr/bin/{{ kolla_container_engine }} stop ovn_controller -t {{ docker_graceful_timeout }}
diff --git a/releasenotes/notes/bug-2093328-db7299bb092fb12b.yaml b/releasenotes/notes/bug-2093328-db7299bb092fb12b.yaml
new file mode 100644
index 000000000..e0045e0ff
--- /dev/null
+++ b/releasenotes/notes/bug-2093328-db7299bb092fb12b.yaml
@@ -0,0 +1,8 @@
+---
+fixes:
+  - |
+    Stopping ovn-controller via systemd will now trigger a graceful shutdown,
+    thus properly deregistering the chassis in OVN.
+    This will result in graceful failover and failback of external gateway
+    logical router ports.
+    `LP#2093328 <https://launchpad.net/bugs/2093328>`__
-- 
2.47.0

